### PR TITLE
PR for Baselines

### PR DESCRIPTION
**Pull request by Micky774**
Do not merge yet.

### CODE REVIEW

File: benchmarks/models.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ def __init__(self):
 
         self.cnn_layers = Sequential(
             # Convolution 1
-            Conv2d(2,2,3,padding=1), # my god I do not know how to select params
+            Conv2d(2,2,3,padding=1),
```
Comment (Micky774) [inline benchmarks/models.py]: We don't need padding unless we're trying to retain the same shape after the convolution layer, which for us is not a priority. Padding introduces zeros at the boundaries which can "dilute" information at the edges in deeper layers, so I prefer to avoid it unless necessary.

File: benchmarks/models.py [hunk 2]
```diff
@@ -30,6 +30,81 @@ def __init__(self):
...
 
+    def forward(self, x):
+        x = self.cnn_layers(x)
+        x = x.view(x.size(0), -1)
+        x = self.linear_layers(x)
+        return x
+
+
+import torch
+from torch.nn import Linear, ReLU, CrossEntropyLoss, Sequential, Conv2d, MaxPool2d, Softmax, Module, BatchNorm2d
```
Comment (Micky774) [inline benchmarks/models.py]:
```suggestion
```

File: benchmarks/models.py [hunk 3]
```diff
@@ -30,6 +30,81 @@ def __init__(self):
...
+
+# very simlar to the first, just expands feature maps immediately and has 2 linear layers
+class Model_1(Module):
+    
+
+    def __init__(self):
+        super(Model_1, self).__init__()
```
Comment (Micky774) [inline benchmarks/models.py]:
```suggestion
        super().__init__()
```

File: benchmarks/models.py [hunk 4]
```diff
@@ -30,6 +30,81 @@ def __init__(self):
...
+        x = self.cnn_layers(x)
+        x = x.view(x.size(0), -1)
+        x = self.linear_layers(x)
+        return x
+
+# Losely styled after vgg model
+class VGG_Model(Module):
+    
+    def __init__(self):
+        super(VGG_Model, self).__init__()
```
Comment (Micky774) [inline benchmarks/models.py]: Same as above
```suggestion
        super().__init__()
```

File: benchmarks/data_utils.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        print(path_to_folder)
+        
+
+        name_constraint = lambda x: 'normalized' in x
+        for label in self.class_types:
+            target =  self.class_types.index(label) #class 0,1,2
+            path = os.path.join(path_to_folder, label)
+            files = os.listdir(path)
+            for file_name in files:
+                    if  name_constraint(file_name):
```
Comment (Micky774) [inline benchmarks/data_utils.py]: We don't need a lambda for something this simple -- just include the literal.

```suggestion
                    if 'normalized' in file_name:
```

File: benchmarks/data_utils.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        self.targets= []
+        self.vid_path = []
+        self.transform = transform
+        # self.transform_input = transform_input
+        self.class_types = class_types
+
+        print(path_to_folder)
+        
+
+        name_constraint = lambda x: 'normalized' in x
```
Comment (Micky774) [inline benchmarks/data_utils.py]:
```suggestion
```

File: benchmarks/data_utils.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        
+    def __len__(self):
+        return len(self.targets)
+
+    
+    def __getitem__(self, idx):
+        target_class = self.targets[idx]
+        vid = np.load(self.vid_path[idx])
+        first_frame = vid[0]
+        last_frame = vid[-1]
```
Comment (Micky774) [inline benchmarks/data_utils.py]: We should consider building a static dataset where we _only_ have the first and last frames instead of selecting them from a complete video numpy array as we do here.

File: benchmarks/data_utils.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        target_class = self.targets[idx]
+        vid = np.load(self.vid_path[idx])
+        first_frame = vid[0]
+        last_frame = vid[-1] 
+
+        first_frame = torch.as_tensor((first_frame))
+        last_frame = torch.as_tensor((last_frame))
+
+        # stack frames as channels, x shape is [2, img_width, img_height]
+        sample = torch.stack((first_frame, last_frame))
```
Comment (Micky774) [inline benchmarks/data_utils.py]: This can be simplified by stacking them as numpy arrays first, then converting them to tensors.

File: benchmarks/data_utils.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        first_frame, last_frame = sample.split(1, dim=0)
+
+
+        if self.use_first_frame:
+            img_blurred = TF.gaussian_blur(first_frame, kernel_size=self.kernel_size) # TODO setting?
+        else:
+            img_blurred = TF.gaussian_blur(last_frame, kernel_size=self.kernel_size) # TODO setting?
+
+        # used the topk function in case we later decide to take multiple peaks
+        values, indices = torch.topk(img_blurred.flatten(), k=1)
```
Comment (Micky774) [inline benchmarks/data_utils.py]: Well done!

File: benchmarks/data_utils.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+        """ 
+        Need to address the empty images (no mitochondria pictured) before train/test stage
+        """
+        # If first frame is empty, check the last frame for a signal
+        if values[0] == 0.0:
+            img_blurred = TF.gaussian_blur(last_frame, kernel_size=self.kernel_size) # TODO setting?
+            values, indices = torch.topk(img_blurred.flatten(), k=1)
+            if values[0] == 0.0:
```
Comment (Micky774) [inline benchmarks/data_utils.py:89]: Can you confirm that this code is actually ever run?

File: benchmarks/data_utils.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        left = int(np_idx[2] - 1/2 * self.height)
+        assert first_frame.shape == img_blurred.shape
+
+        # The crop is taken from original, unblurred image
+        first_cropped = torchvision.transforms.functional.crop(first_frame, top, left, self.width, self.height)
+        last_cropped = torchvision.transforms.functional.crop(last_frame, top, left, self.width, self.height)
+        cropped = torch.cat((first_cropped, last_cropped), 0)
+        # print("DEBUG: roi input size", cropped.size())
+
+        return cropped
```
Comment (Micky774) [inline benchmarks/data_utils.py]:
```suggestion
        return TF.crop(sample, top, left, self.width, self.height)
```

File: benchmarks/data_utils.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+        """ 
+        Need to address the empty images (no mitochondria pictured) before train/test stage
+        """
+        # If first frame is empty, check the last frame for a signal
+        if values[0] == 0.0:
+            img_blurred = TF.gaussian_blur(last_frame, kernel_size=self.kernel_size) # TODO setting?
+            values, indices = torch.topk(img_blurred.flatten(), k=1)
+            if values[0] == 0.0:
+                print("NO DATA: maximum pixel value is {}, but this sample will still be trained on".format(values[0]))
```
Comment (Micky774) [inline benchmarks/data_utils.py]: I prefer using `f-strings` since they're more readable and manageable in most cases, imo.
```suggestion
                print(f"NO DATA: maximum pixel value is {values[0]}, but this sample will still be trained on")
```

File: benchmarks/data_utils.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        self.height = window_size[1]
+        self.kernel_size = kernel_size
+        self.use_first_frame = use_first_frame_values
+
+    def __call__(self, sample):
+        """
+        sample is [n_channels, img_width, img_height]
+        """
+
+        first_frame, last_frame = sample.split(1, dim=0)
```
Comment (Micky774) [inline benchmarks/data_utils.py]: I'm not sure why this would work -- if the incoming data is of shape `[n_channels, img_width, img_height]` then it seems that it's coming in _after_ a conv layer, in which case we can't differentiate any channel as the "first" or "last" frames. 

Instead, perhaps it should be the case that the expected shape is just `[2, img_width, img_height]`? In which case it should suffice to set 
```python
first_frame, last_frame = sample
```

File: benchmarks/data_utils.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        if values[0] == 0.0:
+            img_blurred = TF.gaussian_blur(last_frame, kernel_size=self.kernel_size) # TODO setting?
+            values, indices = torch.topk(img_blurred.flatten(), k=1)
+            if values[0] == 0.0:
+                print("NO DATA: maximum pixel value is {}, but this sample will still be trained on".format(values[0]))
+
+        np_idx = np.array(np.unravel_index(indices.numpy(), img_blurred.shape))
+
+        top = int(np_idx[1] - 1/2 * self.width)
+        left = int(np_idx[2] - 1/2 * self.height)
```
Comment (Micky774) [inline benchmarks/data_utils.py]: I'm not sure why we need this shifting, nor really what it does

File: benchmarks/models.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
+import torch
+from torch.nn import Linear, ReLU, CrossEntropyLoss, Sequential, Conv2d, MaxPool2d, Softmax, Module, BatchNorm2d
+
+
+class Model_0(Module):
```
Comment (Micky774) [inline benchmarks/models.py]: Rename them to be a bit more informative -- e.g. BaseCNN and ExpandedCNN, or really anything that differentiates them.

File: benchmarks/models.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    
+
+    def __init__(self):
+        super(Model_0, self).__init__()
```
Comment (Micky774) [inline benchmarks/models.py]: In python `3.X` you don't need to put anything in `super`
```suggestion
        super().__init__()
```

**Review by Micky774 — COMMENTED** [review]
I think we should also consider using an out-of-the-box ResNet model just to see what kind of results we can achieve with some potent complexity behind the model. 

In general, we should also run/test these idividual functions to make sure they're performing properly. You can do this manually, or if you're interested, I can set up a test suite config and you can then write test cases that will automatically run with a single command.

Looks good so far though, just need some minor changes.

**Comment by Micky774** [comment]
@rmattson1008 Finally got some feedback for you :)

File: benchmarks/data_utils.py (hunk 7 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: That should be pulling from `/data/ornet/single_cell_cnns`, which should be the static dataset you want. Just wrote it that way so it could work on complete videos if need be.

File: benchmarks/data_utils.py (hunk 10 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py:89]: I used to have some print statements that made it clear when the conditional was triggered, (like 4-5 of our samples). I agree with one of your other comments that I should add better testing.

File: benchmarks/data_utils.py (hunk 11 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: This function is set up to take in 2 channel samples and output 2 channel crops. I don't think we want to return just one frame without bigger changes. I couldn't think of an easier way to share the crop coordinates between frames. I definitely should add better documentation.

File: benchmarks/data_utils.py (hunk 12 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: what is this sorcery

File: benchmarks/data_utils.py (hunk 13 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: Yeah, I wanted the results to be [1, width, height]. I think one of the subsequent torch functions needs that format.

File: benchmarks/models.py (hunk 15 above)
Comment (rmattson1008) [reply benchmarks/models.py]: Yeah model 0 and 1 really aren't significantly different - I'm thinking about nixing 0 and having model 1 be our "base". Then we can adjust to no padding.

File: benchmarks/data_utils.py (hunk 10 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py:89]: Yeah if one actually trains on this data, it prints that it can't find any pixel values > 0 for 5 samples. Definitely not the clearest way to handle.

File: benchmarks/data_utils.py (hunk 11 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: I changed this method significantly, the use of tensors might be clearer/better now

File: benchmarks/data_utils.py (hunk 14 above)
Comment (rmattson1008) [reply benchmarks/data_utils.py]: yeah I need to test this.
